---
format:
  html:
    smooth-scroll: true
    highlight-style: github 
    embed-resources: true
    code-overflow: scroll
    code-block-background: true
---

# Schedule

The content of this workshop is based on material produced by The Carpentries, who aim to teach fundamental skills required for conducting research.

Their workshop utilises the command line, whereas today we will be primarily be working with a graphic user interface (GUI).

For more information on the command-line workshop see [Carpentries Metagenomic Workshop](https://carpentries-lab.github.io/metagenomics-analysis/index.html) and [publication in JOSE](https://jose.theoj.org/papers/10.21105/jose.00209).

If you are interested in using the command line for genomic data analysis and are unsure where to start, then I would recommend the lessons by the Data Carpentry Team, [Introduction to the Command Line for Genomics](https://datacarpentry.org/shell-genomics/) and [Data Wrangling and Processing for Genomics](https://datacarpentry.org/wrangling-genomics/).

To make this workshop as accessible as possible, we have opted to use a GUI approach that incorporates many of the data analysis steps outlined in The Carpentries Metagenomic Workshop linked above.

## **Schedule**

|       | Section                                                           | Questions Addressed                                                           |
|-------|-------------------------------------------------------------------|-------------------------------------------------------------------------------|
| 00:00 | 1\. [Starting a Metagenomics Project](01-background-metadata.qmd) | How do you plan a metagenomics experiment?                                    |
| 00:30 | 2\. [Assessing Read Quality](02-assessing-read-quality.qmd)       | How can I described the quality of my data?                                   |
| 01:20 | 3\. [Trimming and Filtering](03-trimming-and-filtering.qmd)       | How can we get rid of sequence data that does not meet our quality standards? |
| 02:15 | 4\. [Metabarcode Assembly](04-assembly.qmd)                       | Why should metabarcode data be merged?                                        |
|       |                                                                   | What is the difference between reads and contigs?                             |
|       |                                                                   | How can we merge metabarcode reads?                                           |
| 03:05 | 5\. [Taxonomic Assignment](05-taxonomic-assignment.qmd)           | How can I know to which taxa my sequences belong?                             |
| 03:55 | 6\. [Exploring Taxonomy with R](06-exploring-taxonomy.qmd)        | How can I use my taxonomic assignment results to analyse?                     |
| 04:20 | 7\. [Diversity Tackled with R](07-tackling-diversity-with-R.qmd)  | How can we measure diversity?                                                 |
|       |                                                                   | How can I use R to analyse diversity?                                         |
| 05:15 | 8\. [Taxonomic Analysis with R](08-abundance-analyses.qmd)        | How can we know which taxa are in our samples?                                |
|       |                                                                   | How can we compare depth-contrasting samples?                                 |
|       |                                                                   | How can we manipulate our data to deliver a message?                          |
| 06:15 | Finish                                                            |                                                                               |

: {.striped .hover}

The actual schedule may vary slightly.
